### PR TITLE
Fix logic for setting keyboard layout

### LINKED ISSUE #62838: [BUG] Setting keyboard layout fails but returns success
**Description**
When the user applies a state to set a keyboard keymap that is not available,
the command fails, but salt reports a success.
I think I have found the cause and am working on a patch.

**Setup**
I think this would affect everyone using this module, but this is where
I encountered it:
- [ ] on-prem machine
- [x] VM (libvirt, qemu, suse SLES15SP4 minion, provisioned with sumaform)
- [ ] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [ ] classic packaging
- [ ] onedir packaging
- [ ] used bootstrap to install

**Steps to Reproduce the behavior**

Run:
`salt-call --local state.single keyboard.system ca.map.gz`
```
[ERROR   ] Command 'localectl' failed with return code: 1
[ERROR   ] stdout: Failed to set keymap: Keymap ca.map.gz is not installed.
[ERROR   ] retcode: 1
[ERROR   ] Command 'localectl' failed with return code: 1
[ERROR   ] output: Failed to set keymap: Keymap ca.map.gz is not installed.
local:
----------
          ID: ca.map.gz
    Function: keyboard.system
      Result: True
     Comment: Set system keyboard layout ca.map.gz
     Started: 17:53:24.783471
    Duration: 214.29 ms
     Changes:   
              ----------
              layout:
                  ca.map.gz

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 214.290 ms
```

**Expected behavior**
Result should be False, Succeeded should be 0, Failed should be 1.

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
```yaml
Salt Version:
          Salt: 3004
 
Dependency Versions:
          cffi: 1.13.2
      cherrypy: Not Installed
      dateutil: Not Installed
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: 0.38.0
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.17
      pycrypto: Not Installed
  pycryptodome: Not Installed
        pygit2: Not Installed
        Python: 3.6.15 (default, Sep 23 2021, 15:41:43) [GCC]
  python-gnupg: Not Installed
        PyYAML: 5.4.1
         PyZMQ: 17.1.2
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.3
 
System Versions:
          dist: sles 15.4 
        locale: UTF-8
       machine: x86_64
       release: 5.14.21-150400.22-default
        system: Linux
       version: SLES 15.4```
 
</details>

**Additional context**
I think I've worked out why this happens and am working on a fix, over here:
https://github.com/saltstack/salt/compare/master...zzaimeche:salt:fix-keyboard-fails
I think Salt does not check whether the shell command was successful, but whether
the python calling the command ran successfully. So whether the shell command
succeeds or fails, Salt reports a success.
The issue is in https://github.com/saltstack/salt/blob/master/salt/modules/keyboard.py
and https://github.com/saltstack/salt/blob/master/salt/states/keyboard.py .
I think Salt needs to check for the retcode, so I used cmd.retcode based on other
examples in the codebase, though I couldn't find the documentation for cmd.retcode,
so maybe there's a better way. My WIP patch seems to fix
the issue on my machine (but my WIP patch will break things on other
distros at the moment). I'll make a PR shortly. I'm hoping someone from the Salt
team can take a look and advise if there's a preferred approach, since I suspect there
might have been similar issues with other modules and there may be a convention for
this kind of fix.

### PR DESCRIPTION
**Pull request by zzaimeche**
Currently, when a command fails to set the keyboard layout, salt reports a success. This commit fixes the logic so that the right message is displayed, by checking the command's return code.

Systems running Debian, RedHat and Gentoo will still have the bug; it would be better to fix those, too, but I'm not sure how to get the retcode for those yet because they set the layout with a different command. The x changes have also not been manually tested though those _should_ work. The system changes do fix the issue on my minion.

### What issues does this PR fix or reference?
Fixes: #62838 

### Previous Behavior
When an unavailable keyboard layout was chosen, Salt would report a success but fail to set it.

### New Behavior
When an unavailable keyboard layout is chosen, Salt reports a failure.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by zzaimeche** [comment]
Tests fail beacuse cmd.retcode isn't recognised. It works on my system and it's used elsewhere in the code, so hoping someone who knows more about it can advise about that.

File: salt/modules/keyboard.py
Comment (Ch3LL) [inline salt/modules/keyboard.py:76]: I see your TODO comment, but I would prefer to not merge this without getting the result of thise command and the other commands per each OS.

File: salt/modules/keyboard.py
Comment (Ch3LL) [inline salt/modules/keyboard.py:87]: This is changing behavior. If the commands where a success we should still return `layout` which is what the previous behavior was.

File: salt/modules/keyboard.py
Comment (Ch3LL) [inline salt/modules/keyboard.py:116]: Same here, we are changing behavior.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/keyboard.py
Comment (zzaimeche) [reply salt/modules/keyboard.py:87]: Ah sorry, yeah, it does change what's returned from that function and might break things for someone who's using the module function directly. There's nothing I could find that does in the salt codebase, so I'm not sure if that's likely, but someone might have built their own tools that would be affected.

What I mean is that from the user's perspective, it doesn't change what happens when the user runs the command to set the keyboard layout. If I kept it returning layout, the calling function would now fail, because the calling function now expects this function to return a return code. It's confusing to write out but I ended up with this:

Previously the state function called the module function and gave it a layout. if the module function finished running, it would still have something saved as 'layout', so it would give that back to the state function. the state function would check it got anything at all back from the module function, and if it did, it would report success. It doesn't care if the layout is actually set, or if the layout variable contains anything valid; it just has to contain *something*. So in practice, this just becomes a check for if the function runs all the way without erroring.

So although it returns 'layout', it doesn't do anything with it and you could use any variable at all that the calling function provided. (Or check for a return code directly.)

I put these here so that if an end user on gentoo/redhat/debian ran the set-keyboard commands, they wouldn't see any difference in behaviour to before, instead of breaking it for them. But it would be better to actually _fix_ it for them. :) I think the correct behaviour is for all of them to return 0 iff the layout is set.

Comment (Ch3LL) [reply salt/modules/keyboard.py:87]: I agree there is not a alot of value in returning `layout`, but we should still properly deprecate that the return of this function is going to change.

And instead of returning 0, would it make more sense to return True vs False instead, which is inline with a lot of other salt functions.

Let me get some more memebers of my team's opinions on the deprecating the return and what we should return.

@whytewolf @garethgreenaway thoughts here?

Comment (zzaimeche) [reply salt/modules/keyboard.py:87]: Thank you both for the feedback! Okay, I'll try to fix that.

Comment (zzaimeche) [reply salt/modules/keyboard.py:90]: Thanks again for your help so far! :) I've adjusted it to return True instead of 0, and updated the unit tests to match. I haven't yet been able to test the behaviour when localectl is unavailable (I currently expect I'll be able to do that for redhat and debian, but it may be harder for gentoo). I'm also not sure what I need to do in order to follow the deprecation path with the return. I found https://docs.saltproject.io/en/latest/topics/development/deprecations.html ; does this mean I should make a separate patch to merge first, that will just contain the warning?

Comment (Ch3LL) [reply salt/modules/keyboard.py:90]: Normally we add a warning in the function that will spit out a deprecation message. But I think this is going to be too noisy since we are just letting users now that the return of this function will be changing in an upcoming release.

How about just adding an additional changelog file with the file extension `.deprecated` and include a message that
the `set_sys` and `get_x` functions in the keyboard module will change in the Argon release to return True instead of the value of `layout`.

Then we should be good to go forward with PR.

Comment (Ch3LL) [reply salt/modules/keyboard.py:90]: @zzaimeche just making sure you saw my feedback ^

Comment (zzaimeche) [reply salt/modules/keyboard.py:90]: @Ch3LL Thanks! I can add that file, then. :)

File: salt/modules/keyboard.py
Comment (Ch3LL) [inline salt/modules/keyboard.py:90]: Sorry I missed this in my previous review. We still want to ensure we return `layout` for both this function and the `get_x` function until Argon

Can you also add a comment in both functions:

```
# salt.utils.versions.warn_until(Argon, "The set_sys function will return True instead of layout in the Argon release)
```

```
# salt.utils.versions.warn_until(Argon, "The get_x function will return True instead of layout in the Argon release)
```

The reason being thats how we detect deprecations for a given release and we can then change the return to True instead of layout in that release.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Comment by zzaimeche** [comment]
I've made a patch for deprecations, https://github.com/saltstack/salt/pull/63075 ; I've removed the deprecation entry from the changelog in this one and put it in there along with the comments.

**Comment by Ch3LL** [comment]
It looks like we are still returning True/False here. I'm not entirely clear of the purpose of the other PR as we don't want to warn every time someone runs those commands. Taht would be too noisy. Having the changelog to bring attention to users that it will be deprecated will be enough. But we need to ensure we are still returning layout in this PR before we can merge.

**Comment by Ch3LL** [comment]
oh and add the deprecation changelog file on this PR as well.

**Comment by Ch3LL** [comment]
Okay, that makes sense. Yeah we will want those warnings to be comments, just so we can find it and deprecate it.  I'll comment on the PR

We can't actually make the change until before Argon is released. So the best approach would be to jsut update this PR that states if True return layout. And then in the Argon release someone can update the module to return True/False.

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**
